### PR TITLE
Raised NotImplementedError for base classes without their own impleme…

### LINKED ISSUE #3294: Add `NotImplementedError` for functions that'll be overridden in Python
### Description of the task

To prevent functions that we do not expect to be overridden to be executed in Python, we should raise a `NotImplementedError`

Here are the places (that I know about) where we should add this:
- [x] https://github.com/UBC-Thunderbots/Software/blob/3cc5a2d4c45205f3b0dc7eae6e85f328e2973160/src/software/thunderscope/gl/widgets/gl_toolbar.py#L31

### Acceptance criteria

- [x] Replace known instances of these abstract function with an implementation that raises a `NotImplementedError`

### PR DESCRIPTION
**Pull request by annieisawesome2**
### Description

Tagged necessary super-classes with abstract method so not implemented error is raised when base class did not implement the method. I went ahead and added this to all the super classes that had methods that needed to be implemented by child classes. (gl_toolbar, tbots_test_runner, gl_shape, gl_layer, validation, time_provider. )

 
### Testing Done

Commented out implementation of child class and ran the build to see if the not implemented error and message would be raised. 

### Resolved Issues
resolves #3294

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ x] **Remove all commented out code**
- [ x] **Remove extra print statements**: for example, those just used for testing
- [x ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Review by williamckha — COMMENTED** [review]
Awesome work! One small thing -- in order to use the `@abstractmethod` decorator, you need the class to inherit the [ABC](https://docs.python.org/3/library/abc.html#abc.ABC) class

**Comment by annieisawesome2** [comment]
> Awesome work! One small thing -- in order to use the `@abstractmethod` decorator, you need the class to inherit the [ABC](https://docs.python.org/3/library/abc.html#abc.ABC) class

Sounds good! I'll definitely take note of that for next time, thank you :)

**Comment by Andrewyx** [comment]
I think what william is talking about is this: https://docs.python.org/3/library/abc.html#abc.ABC:~:text=The%20abc%20module,derived%20from%20it.

**Comment by williamckha** [comment]
You can update this PR by committing to your `annie/abstract_method_pr` branch, and then do `git push` (or just use the CLion UI to push). 

Before pushing, you will need to `git pull upstream master` because Arun's Ubuntu on ARM PR got merged into master. This PR added some new [CI checks](https://github.com/user-attachments/assets/57482bee-0106-4cf3-b5de-5102d08bd84d) that need to pass in order for you to merge your PR into master. You will also probably need to do `git pull origin annie/abstract_method_pr` before pushing because our auto-fixes bot committed some changes to your branch that you won't have on your local machine.

**Comment by annieisawesome2** [comment]
> You can update this PR by committing to your `annie/abstract_method_pr` branch, and then do `git push` (or just use the CLion UI to push).
> 
> Before pushing, you will need to `git pull upstream master` because Arun's Ubuntu on ARM PR got merged into master. This PR added some new [CI checks](https://github.com/user-attachments/assets/57482bee-0106-4cf3-b5de-5102d08bd84d) that need to pass in order for you to merge your PR into master. You will also probably need to do `git pull origin annie/abstract_method_pr` before pushing because our auto-fixes bot committed some changes to your branch that you won't have on your local machine.

Alright I'll take a look at updating with the new checks before sending in another PR for this issue, thanks!

File: src/software/thunderscope/gl/layers/gl_layer.py [hunk 1]
```diff
@@ -35,6 +35,7 @@ def __init__(self, name: str) -> None:
```
Comment (someone2060) [inline src/software/thunderscope/gl/layers/gl_layer.py:35]: Not sure whether `@abstractmethod` is needed here

**Review by someone2060 — COMMENTED** [review]
Generally, this looks good to me!

I noticed an inconsistency in `gl_layer.py`, but I'm unsure if `GLLayer` can be abstract. Could other reviewers look over this file and give their thoughts?

File: src/software/thunderscope/gl/layers/gl_layer.py (hunk 1 above)
Comment (williamckha) [reply src/software/thunderscope/gl/layers/gl_layer.py:35]: We actually removed `ABC` and `@abstractmethod` from a couple classes last saturday bc it was causing issues

**Comment by williamckha** [comment]
I've found a way to fix the problem we were running into with the classes that inherit both a PyQt class (i.e. something that derives from `QObject`) and `ABC`. Basically, `QObject` and `ABC` have different metaclasses (some strange Python thing) and so Python doesn't know which should be the metaclass for the class that inherits both `QObject` and `ABC`. We can get around this by making a new metaclass that inherits the metaclasses of `QObject` and `ABC`:

```python
# Create a new metaclass
class QObjectABCMeta(type(QtCore.QObject), abc.ABCMeta):
    pass

# Explicitly set the metaclass of GLShape to our new QObjectABCMeta
class GLShape(GLLinePlotItem, ABC, metaclass=QObjectABCMeta):
    ...
```

This is rather confusing, so I'll leave it up to you whether you want to do this or just not use `ABC` for `QObject` subclasses.

**Comment by annieisawesome2** [comment]
> I've found a way to fix the problem we were running into with the classes that inherit both a PyQt class (i.e. something that derives from `QObject`) and `ABC`. Basically, `QObject` and `ABC` have different metaclasses (some strange Python thing) and so Python doesn't know which should be the metaclass for the class that inherits both `QObject` and `ABC`. We can get around this by making a new metaclass that inherits the metaclasses of `QObject` and `ABC`:
> 
> ```python
> # Create a new metaclass
> class QObjectABCMeta(type(QtCore.QObject), abc.ABCMeta):
>     pass
> 
> # Explicitly set the metaclass of GLShape to our new QObjectABCMeta
> class GLShape(GLLinePlotItem, ABC, metaclass=QObjectABCMeta):
>     ...
> ```
> 
> This is rather confusing, so I'll leave it up to you whether you want to do this or just not use `ABC` for `QObject` subclasses.

sounds good! Ill definitely look into QObject sometime!

**Review by itsarune — APPROVED** [review]
Awesome work!

**Review by someone2060 — APPROVED** [review]
Looks good to me!

**Review by Mr-Anyone — APPROVED** [review]
nice

**Pull request merged**